### PR TITLE
RAJA launch API updates

### PR DESCRIPTION
**Pull request by artv3**
The API formally known as Teams is now RAJA Launch and it has a few API changes.

TODO:
- [x] Wait for the next RAJA release, v2022.10.3.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3293](https://github.com/mfem/mfem/pull/3293) | @artv3 | @tzanio | @v-dobrev + @tzanio + @camierjs | 11/1/22 | 12/12/22 | 1/3/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by artv3** [comment]
There is a RAJA release coming out soon to fix a bug, we should wait for it before merging.

File: CHANGELOG
Comment (artv3) [inline CHANGELOG:264]: RAJA 2022.10.2 will be out soon.

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: general/forall.hpp
Comment (tzanio) [inline general/forall.hpp:119]: Do we want to maintain the support for older versions of RAJA?

Comment (tzanio) [reply general/forall.hpp:119]: ping: @artv3, @kennyweiss ☝️

Comment (artv3) [reply general/forall.hpp:119]: I chatted with @kennyweiss, and since we are moving RAJA::launch out of expt it may make sense to encourage people to move to the more stable API now. In addition its easier to maintain one API.

Comment (tzanio) [reply general/forall.hpp:119]: thanks for clarifying this

**Review by tzanio — APPROVED** [review]

**Comment by tzanio** [comment]
@v-dobrev and @camierjs, can you please review when you get a chance?

File: CHANGELOG
Comment (artv3) [reply CHANGELOG:264]: RAJA 2022.10.2 was released a few days ago.

**Review by camierjs — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by camierjs** [comment]
@artv3 , we might need the `-lcamp` library to be added too.

**Comment by tzanio** [comment]
The merge in `next` was reverted so we should be careful when we merge this next time.

File: config/defaults.mk
Comment (artv3) [inline config/defaults.mk:518]: @camierjs @v-dobrev , if both umpire and raja need -lcamp is this right right way to add the flag?

Comment (v-dobrev) [reply config/defaults.mk:518]: When `CAMP_DIR` is defined (i.e. RAJA/UMPIRE use external CAMP) then we need something like this before `-lcamp`: `$(XLINKER)-rpath,$(CAMP_DIR)/lib -L$(CAMP_DIR)/lib`.

It is probably best to define
```
CAMP_LIB = -lcamp
ifdef CAMP_DIR
   CAMP_LIB = $(XLINKER)-rpath,$(CAMP_DIR)/lib -L$(CAMP_DIR)/lib -lcamp
endif
```
and then use just `$(CAMP_LIB)` in `RAJA_LIB` and `UMPIRE_LIB`.

**Comment by v-dobrev** [comment]
I tried the CMake build system and saw that it does not import the Camp dependency automatically. Before, this was not an issue because the default RAJA build (with internal Camp) did not require additional flags. However, now we need at least to add `-lcamp`.

I tried updating our `FindRAJA.cmake` file with the new path to `raja-config.cmake` here: https://github.com/mfem/mfem/blob/b09cc38df9f1da182a5625c92660737ca02e710a/config/cmake/modules/FindRAJA.cmake#L22
with `"${RAJA_DIR}/lib/cmake/raja/raja-config.cmake"`, however, that has a problem: I get this error:
```
CMake Error at /<deleted>/share/cmake-3.24/Modules/CMakeFindDependencyMacro.cmake:47 (find_package):
  Found package configuration file:

    /<deleted>/raja/lib/cmake/camp/campConfig.cmake

  but it set camp_FOUND to FALSE so package "camp" is considered to be NOT
  FOUND.  Reason given by package:

  The following imported targets are referenced, but are missing:
  blt::blt_hip_runtime

Call Stack (most recent call first):
  /<deleted>/raja/lib/cmake/raja/raja-config.cmake:73 (find_dependency)
  config/cmake/modules/FindRAJA.cmake:43 (include)
  CMakeLists.txt:425 (find_package)
```
I also tried importing that config in the more standard way using:
```cmake
   find_package(RAJA REQUIRED CONFIG NO_DEFAULT_PATH PATHS "${RAJA_DIR}")
```
but that gives the same exact error.

**Comment by v-dobrev** [comment]
I pushed an update and fix for the CMake issue above.

I used a workaround for the error posted above which should really be fixed in RAJA: the inclusion of `bltTargets.cmake` from `raja-config.cmake` should be done before the call to `find_dependency(camp ...)`.

**Comment by artv3** [comment]
> I pushed an update and fix for the CMake issue above.
> 
> I used a workaround for the error posted above which should really be fixed in RAJA: the inclusion of `bltTargets.cmake` from `raja-config.cmake` should be done before the call to `find_dependency(camp ...)`.

@v-dobrev , which version of RAJA are you using? It sounds like we may have fixed that already, cc @davidbeckingsale

**Comment by artv3** [comment]
> > I pushed an update and fix for the CMake issue above.
> > I used a workaround for the error posted above which should really be fixed in RAJA: the inclusion of `bltTargets.cmake` from `raja-config.cmake` should be done before the call to `find_dependency(camp ...)`.
> 
> @v-dobrev , which version of RAJA are you using? It sounds like we may have fixed that already, cc @davidbeckingsale

Can you try the develop version of RAJA?

**Comment by v-dobrev** [comment]
@artv3, I was using the latest RAJA `develop` branch as of Friday (think). The specific RAJA hash is https://github.com/LLNL/RAJA/commits/6a875c2d436481601c556af148e425b4a1cfe967.

File: config/defaults.mk
Comment (artv3) [reply config/defaults.mk]: I just pushed up the adjustments

**Review by v-dobrev — APPROVED** [review]
We should wait for the next RAJA release before merging this.

**Comment by artv3** [comment]
> We should wait for the next RAJA release before merging this.

We have released the code! 
https://github.com/LLNL/RAJA/releases/tag/v2022.10.3

File: config/cmake/modules/FindRAJA.cmake
Comment (v-dobrev) [inline config/cmake/modules/FindRAJA.cmake:30]: With the new 2022.10.3 release of RAJA we can probably remove this workaround but we should test without it to make sure it is no longer necessary.

Comment (v-dobrev) [reply config/cmake/modules/FindRAJA.cmake]: Note: when testing with HIP, I had to use the fix from #3347.

**Comment by tzanio** [comment]
Are we ready to re-merge this in `next`?

(this will not be trivial because of the revert commit https://github.com/mfem/mfem/pull/3293#issuecomment-1314221982

**Comment by v-dobrev** [comment]
> Are we ready to re-merge this in `next`?

Yes, I think we are.

**Comment by tzanio** [comment]
Merged in `next` for testing...

File: CMakeLists.txt
Comment (v-dobrev) [inline CMakeLists.txt:90]: This `if` statement can be merged with the above `if` statement.

**Pull request merged**
